Flexible indexes refactoring notes (#4979)
pydata/xarray: FLEXIBLE_INDEXES_NOTES.md

# Proposal: Xarray flexible indexes refactoring

Current status: https://github.com/pydata/xarray/projects/1

TODO: check if the project is up-to-date and update it if necessary.

## 1. Data Model

Indexes are used in Xarray to extract data from Xarray objects using coordinate labels instead of using integer array indices. Although the indexes used in an Xarray object can be accessed (or built on-the-fly) via public methods like `to_index()` or properties like `indexes`, those are mainly used internally.

The goal of this project is to make those indexes 1st-class citizens of Xarray's data model. As such, indexes should clearly be separated from Xarray coordinates with the following relationships:

- Index -> Coordinate: one-to-many
- Coordinate -> Index: one-to-zero-or-one

> An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.

Reviewer: I think this is a reasonable idea, but perhaps could be implemented via a "meta" index that wraps multiple other indexes internally.

`method='nearest'` in `pd.Index` is already sort of like this: it uses `np.searchsorted()` rather than the hashtable used for other look-ups.

The advantage of only associating one index with a coordinate is that looking up which index to use from an indexing operation is unambiguous and fast (using a dict).

> An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.

Author: Yes I agree that is a relevant example. Another use case could be some tree-based indexes that only support point-wise indexing while others would allow providing slices to `.sel()` (range-based queries of irregular data).

I like the "meta" index idea. It would prevent having to handle complicated situations in Xarray such as checking for potential ambiguous situations where two indexes assigned to the same coordinate both support the same operations. It would also keep the API simple and less confusing for users.

The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.

Xarray's current data model is already based on the same index-coordinate relationships. The current data model also implies a one-to-one relationship between a dimension and an index. This one-to-one relationship works as currently it is not possible to perform label-based data selection in Xarray using multi-dimensional coordinates, but since we want to enable this feature in this proposal, we need the dimension-index relationship to evolve towards many-to-many.

In the example below, we'd like to select data points based on their x, y values and/or on their latitude/longitude positions:

```
>>> da
<xarray.DataArray (x: 2, y: 2)>
array([[5.4, 7.8],
       [6.2, 4.7]])
Coordinates:
  * lon      (x, y) float64 10.2 15.2 12.6 17.6
  * lat      (x, y) float64 40.2 45.6 42.2 47.6
  * x        (x) float64 200.0 400.0
  * y        (y) float64 800.0 1e+03

> >>> da.sel(x=..., y=..., lon=..., lat=...)

Reviewer: This may not be the best example, because I suspect this particular case should be an error because lat/lon overlap with x/y. The desired behavior is ambiguous.

E.g., `da.sel(lon=..., lat=...)` and `ds.sel(x=..., y=...)` each should be fine but not the combination.

Author: Yes indeed. I was wondering how we should handle this:

- A. do not allow any combination
- B. add a new option in `.sel()` to specify how the query results from each index should be combined (i.e., union vs. intersection, etc.)

```

We would need one geographic index for the `lat` and `lon` coordinates and two indexes for the `x` and `y` coordinates, respectively.

## 2. Proposed API changes

### 2.1 Index wrapper classes

Every index that is used to select data from Xarray objects should inherit from a base class, e.g., `XarrayIndex`, that provides some common API. `XarrayIndex` subclasses would generally consist of thin wrappers around existing index classes such as `pandas.Index`, `pandas.MultiIndex`, `scipy.spatial.KDTree`, etc.

There is a variety of features that an xarray index wrapper may or may not support:

- 1-dimensional vs. 2-dimensional vs. n-dimensional coordinate
- built from a single vs multiple coordinate(s)
- in-memory vs. out-of-core (dask) index data/coordinates (vs. other array backends)

> - orthogonal vs. vectorized indexing, range-based vs. point-wise selection

Reviewer: This is for integer based indexing?

Author: Also for label-based indexing. For example, [pys2index](https://github.com/benbovy/pys2index) currently only supports point-wise selection (nearest-neighbors), but it could later wrap other features available in `s2geometry` like selecting all points within latitude and/or longitude range(s).

- exact vs. inexact lookups

Every `XarrayIndex` subclass must at least implement a method that takes label-based indexers as argument and that returns the corresponding position-based indexers. Whether or not a `XarrayIndex` subclass supports each of the features listed above should be either declared explicitly via a common API or left to the implementation.

There are potentially other methods that an `XarrayIndex` subclass must/should/may implement, like Xarray coordinate getters (see [Section 2.2.4](#224_Implicit_coodinates)) or `serialize()`/`deserialize()` methods. `XarrayIndex` API may be expanded in the future.

Xarray should provide a minimal set of built-in index wrappers (this could be reduced to the indexes currently supported in Xarray, i.e., `pandas.Index` and `pandas.MultiIndex`). Other index wrappers may be implemented in 3rd-party libraries (recommended). The `XarrayIndex` base class should be part of Xarray's public API.

#### 2.1.1 Index discoverability

> For better discoverability of Xarray-compatible indexes, Xarray could provide some mechanism to register new index wrappers, e.g., something like [xoak's `IndexRegistry`](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.IndexRegistry.html#xoak.IndexRegistry).

Author: Agreed. That would be consistent with the storage backends and possibly other backends too (#3640). Perhaps it would be nice if somehow we could still quickly and dynamically register indexes, like we do with Dataset/DataArray accessors. I haven't looked yet if the storage backend system allows that.

I think that users should be able to provide `XarrayIndex` subclasses directly in `set_index`, etc. anyway.

### 2.2 Explicit vs. implicit index creation

#### 2.2.1 Dataset/DataArray's `set_index` method

New indexes can be built from an existing set of coordinates or variables in a Dataset/DataArray using the `.set_index()` method.

TODO: describe API updates needed to provide the kind of index that we want to build from a set of coordinates.

#### 2.2.2 Dataset/DataArray's `indexes` constructor argument

TODO

#### 2.2.3 Implicit default indexes

In general explicit index creation should be preferred over implicit index creation. However, there is a majority of cases where basic `pandas.Index` objects could be built and used as indexes for 1-dimensional coordinates. For convenience, Xarray should automatically build such indexes for the coordinates where no index has been explicitly assigned in the Dataset/DataArray constructor or when indexes have been reset / dropped.

For which coordinates?

- A. only 1D coordinates with a name matching their dimension name
- B. all 1D coordinates

When to create it?

- A. each time when a new Dataset/DataArray is created

> - B. only when we need it (i.e., when calling `.sel()` or `indexes`)

Reviewer: I would lean towards (A) and (A).

One related challenge is that indexes need to correspond to _immutable_ coordinate values. Otherwise the index can be invalidated by mutation.

Author: Yes, I think options A and A are what Xarray currently does and there's no real need to change that IMHO.

#### 2.2.4 Implicit coordinates

> If we want to strictly follow the data model defined in [Section 1](#1_Data_Model), we need to ensure that at least one coordinate exists for each index (generally we need to ensure that the data indexed is properly exposed as one or more coordinates), especially when index objects are given directly to the Dataset/DataArray constructor ([Section 2.2.2](#222_DatasetDataArrays_indexes_constructor_argument)). This means that every `XarrayIndex` subclass must implement a method to get or generate one or more Xarray coordinates.

Reviewer: Another option would be to only validate that the corresponding coordinate exist when `indexes` are passed, and not to do implicit creation of coordinates from an index. I think this is viable given that we currently don't support explicitly passing an index.

Author: Hmm currently we can explicitly pass a multi-index as a coordinate:

```python
>>> midx = pd.MultiIndex.from_arrays([['a', 'b'], [0, 1]], names=['lvl1', 'lvl2'])
>>> xr.DataArray([1.0, 2.0], dims='x', coords={'x': midx})
```

Virtual coordinates are created for each level of the multi-index. How should we handle this case after the refactoring? We could silently ignore the levels and simply consider `midx` as an array (i.e., `midx.to_numpy()`). Or we could raise an error...

More generally, I thought that explicitly passing index objects could still be useful since those objects may be costly to build, but I need to think more about whether it's worth the trouble.

Author: > Yes, I would still like to support this syntax when a MultiIndex is passed in coords, but maybe by unpacking as real rather than "virtual" coordinates. This could directly make coordinates with the names lvl1 and lvl2.

Hmm shouldn't we take the opportunity of this refactoring to change this behavior? This looks already pretty confusing to me, and it will be even more if it creates new variable objects that do not have the name passed as key to `coords`. Even though it would be a breaking change, a more predictable behavior would be that any value in the sequence or mapping passed to `coords` would be treated as an array that can be wrapped into an xarray variable, i.e.,

```python
>>> xr.DataArray([1.0, 2.0], dims='x', coords={'x': midx})
<xarray.DataArray (x: 2)>
array([1., 2.])
Coordinates:
    x        (x) object ('a', 0) ('b', 1)
```

If we really want to reuse an existing index object in a Dataset/DataArray (which I assume is not a *very* common case), we could maybe do so via some helper function, e.g.,

```python
>>> coords_dict, index_dict = create_coords_from_index(midx, dims='x')

>>> xr.DataArray([1.0, 2.0], dims='x', coords=coords_dict, indexes=index_dict)
<xarray.DataArray (x: 2)>
array([1., 2.])
Coordinates:
  * lvl1     (x) object 'a' 'b'
  * lvl2     (x) int64 0 1
Indexes:
    lvl1, lvl2   <MultiIndex coords=(lvl1, lvl2) dims=x>
```

> Additionally, we might want to generate a coordinate that is specific to the index, generally with a composite data type, like a coordinate with n-elements tuple labels for a n-levels `pandas.MultiIndex`. This should be optional, though. An `XarrayIndex` may or may not support this. This coordinate probably shouldn't be created implicitly or by default when setting the new index. Also, this might raise some issues, e.g., what would be the name of this coordinate? For a `pandas.MultiIndex` it would make sense in many (but not all) cases to choose the dimension name as the coordinate name.

Reviewer: This is what we do right now, but I'm not sure it's really a good idea.

In the MultiIndex case, I would lean towards a single XarrayIndex object holding the MultiIndex, and a number of 1D coordinate variables for the levels of the MultiIndex.

Author: I agree, I'm not sure either it's a good idea. I haven't had any use case for this, but maybe other users have some? We could still manually create new coordinates from an index object, though.

I'm wondering if this would be temporarily useful for backward compatibility. For example, if we want to still support `da.set_index(dim_x=['coord_a', 'coord_b'])`, would it be ok not having any `dim_x` coordinate in the resulting DataArray?

### 2.3 Index access

#### 2.3.1 Dataset/DataArray's `indexes` property

> The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).

Reviewer: What do we need multiple names in the key for? Are there cases where we want to look-up different indexes depending on which variables are being used?

Rather than allowing for tuple keys, I was imagining we could allow for multiple coordinate names to point to the same index. This would support mapping multiple coordinates to the same (typically multi-dimensional) index, e.g., both `'lat'` and `'lon'` would map to the same index. I think this would align well with the "one-to-zero-or-one" relationship of "Coordinate -> Index" that you describe above.

> The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).

Author: > Right now, it is acceptable to have a tuple as dimension name, so this would be a breaking change

Note that the keys of the mapping returned by the `indexes` property would not correspond here to dimension names but rather one or more coordinate names.

>  Rather than allowing for tuple keys, I was imagining we could allow for multiple coordinate names to point to the same index.

Yes I saw that this is what you suggest in https://github.com/pydata/xarray/pull/2195#issuecomment-442554179. However, I was rather thinking here about something that we could use to quickly iterate over all the indexes in a Dataset/DataArray. Having a collection of unique index objects as values seems better adapted for this use case IMO. It would also align well with the API of `set_index` or the DataArray/Dataset constructors (the `indexes` argument) where we'll probably need to provide some coordinate/variable name(s) -> index class/object as keys -> values. Either way, the structure returned by `indexes` should be consistent with the DataArray/Dataset repr depending on how we show the indexes (I'm not sure that the best repr would be to show an index section with duplicated lines for multi-coordinate indexes, although I don't have a clear idea / strong opinion on this yet).

With what you suggest in https://github.com/pydata/xarray/pull/4979#discussion_r585197016, we could even imagine returning a list of `XarrayIndex` objects, but I think returning a mapping is slightly more convenient as it doesn't hide the index / coordinate(s) relationship behind the index object.

That said, I agree that we need something for getting an index given some coordinate. This is why I suggest another property or method in Section 2.3.2.

> The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).

Reviewer: For the repr, we could probably consolidate entries that map to the same index object, e.g.,
```
Indexes:
  lat, lon     <SpatialIndex coords=(lat, lon) dims=(x, y)>
  x            Float64Index
  y            Float64Index
```

Given that the `repr` prints all the indexes, it is probably OK to group them even if the indexes map has separate entries for `lat` and `lon`. There would just be an extra validation step when you add/remove an index that verifies that the index is consistent on all matching coordinates.

It might be worth sketching out a brief list of trade-offs with each representation for `indexes`, e.g.,
- `Dict[CoordinateName, XarrayIndex]` with duplicated values: Easy to look up all matching indexes by coordinate name, even if the index is over multiple coordinates. Iterating over _unique_ indexes requires some care (e.g., a separate `set` to keep track of Index `id()`).
- `Dict[Tuple[CoordinateName, ...], XarrayIndex]` with unique values: Easy to look-up matching indexes if all coordinate names are supplied in the correct order (but I think this is relatively rare). Cases like `da.sel(lon=..., lat=...)` (shared index, wrong order) or `da.sel(x=..., time=...)` (separate indexes) would require iterating over all indexes and building some sort of temporary structure. Also, there's potential ambiguity between names and tuples of names (if we also allow raw coordinate names in the dict).

> The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).

Author: I think that you convinced me that there's probably more practical reasons to choose `Dict[CoordinateName, XarrayIndex]` as return type for `indexes`.

#### 2.3.2 Additional Dataset/DataArray's convenient methods

> Using tuples as keys in the `indexes` property has the advantage of providing clear information about the order that was used to build multi-coordinate indexes. For convenience, we could have an additional property `coord_indexes` or an additional method `get_index(coord_name: str)` to easily retrieve the index object that is tied to a specific coordinate.

Reviewer: Could this information also live on the `XarrayIndex` object? For example, `MultiIndex` has a `names` attribute.

Author: Yes, and we'll probably need more information. For example, in [xoak](https://github.com/ESM-VFC/xoak/blob/25f3d361e55954f56d95ece9a669f629cf1f01c6/src/xoak/accessor.py#L200-L225) we need the shapes and the dimension names of the coordinates so that we can reformat the query results from the wrapped index (e.g., a 2-d array of shape (n_points, n_coordinates)) back into the format of the indexers given in `.sel()`.

One challenge is to keep this information consistent with the DataArray/Dataset object in which the index is used, as you point out in https://github.com/pydata/xarray/pull/4979#discussion_r585198792.

### 2.4 Propagate indexes through operations

TODO

> - Any operation affecting one coordinate of a multi-coordinate index -> simply drop the index or maybe allow `XarrayIndex` subclasses to optionally generate and return a new or updated index

Reviewer: For MultiIndex, I think of methods like `get_loc_level` that return a new Index when doing indexing.

Author: Yes, currently new indexes may be returned from the [remap_label_indexers](https://github.com/pydata/xarray/blob/48378c4b11c5c2672ff91396d4284743165b4fbe/xarray/core/indexing.py#L243-L275) function, which I think can be replaced by an equivalent (abstract) function in `XarrayIndex` with the same (or a very close) signature.

Since indexes will be loosely coupled to coordinates, we'll likely need to separate returning new indexes from the indexing operation, though.

### 2.5 Using indexes for data selection

TODO: `.sel()` would now accept coordinate names as indexer keys.

> Indexes that are wrapped as `XarrayIndex` subclasses may provide more selection capabilities than what is currently possible using Dataset/DataArray `sel()`, e.g., radius or region selection, k-nearest neighbors, etc. There's no plan here to expand Xarray's selection API. This could be extended in 3rd-party libraries, e.g., using Dataset/DataArray accessors that would reuse the indexes attached to a Dataset/DataArray object.

Author: Yes we would probably need to expose something like `index_kwargs`. Actually, I'm wondering if `method` / `tolerance` should eventually be part of `index_kwargs` as not all indexes might support it.

The examples that I've given here (radius or region selection, k-nearest neighbors) may not be good examples to illustrate my point. I was rather thinking about possible cases where we would like to reuse the indexes to perform specific query operations different than remap label-based indexers to position-based indexers and/or that would require indexer arguments that are not supported in `.sel()` (see, e.g., https://github.com/ESM-VFC/xoak/issues/34).

> Indexes that are wrapped as `XarrayIndex` subclasses may provide more selection capabilities than what is currently possible using Dataset/DataArray `sel()`, e.g., radius or region selection, k-nearest neighbors, etc. There's no plan here to expand Xarray's selection API. This could be extended in 3rd-party libraries, e.g., using Dataset/DataArray accessors that would reuse the indexes attached to a Dataset/DataArray object.

Author: `index_kwargs` may be tricky, though, since the indexers given in `.sel()` may involve different indexes and each index may have its own specific options.

### 2.6 Index serialization

> TODO

Reviewer: I don't think we need serialization for the first proof of concept. Given that we'll force all coordinates corresponding to an index to be explicit variables, direct serialization support may not be necessary. You can simply recreate the index later if necessary.

Author: Agreed. It's certainly too early to address this.

I was thinking about some cases where the index may be very costly to build. We could take advantage of some index implementations to "instantly" re-create the index from its serialized internal data structures (e.g., spatialpandas' [HilbertRTree](https://github.com/holoviz/spatialpandas/blob/4d679f97d750c6a69b49143c1a21e17bf6731d5d/spatialpandas/spatialindex/rtree.py#L68), scikit-learn's `KDTree` and `BallTree` that use an array-based representation of an immutable tree structure). This improvement in performance is welcome but certainly not critical at this stage.

### 2.3 Index access

#### 2.3.1 Dataset/DataArray's `indexes` property

The `indexes` property would return a mapping, where keys are coordinate name(s) (i.e., either a string or a tuple) and values are index objects (i.e., instances of `XarrayIndex`). If a tuple is used as key, the order of its elements should be consistent with the order in which each coordinate participate to the index (see [Section 1](#1_Data_Model)).

#### 2.3.2 Additional Dataset/DataArray's convenient methods

Using tuples as keys in the `indexes` property has the advantage of providing clear information about the order that was used to build multi-coordinate indexes. For convenience, we could have an additional property `coord_indexes` or an additional method `get_index(coord_name: str)` to easily retrieve the index object that is tied to a specific coordinate.

### 2.4 Propagate indexes through operations

TODO

- Any operation affecting one coordinate of a multi-coordinate index -> simply drop the index or maybe allow `XarrayIndex` subclasses to optionally generate and return a new or updated index

> ### 2.5 Using indexes for data selection

Reviewer 2: We should also discuss "indexes for alignment" but maybe now is too early.

For e.g. a proposed CRSIndex (https://github.com/pydata/xarray/issues/2996) would refuse to align except in cases of exact equality.

Or a user might ask `Float64Index` indexes to align to within some specified tolerance: https://github.com/pydata/xarray/pull/4489

Author: I don't think it's too early to discuss this (I haven't added yet a section in the notes). It's an important aspect to consider (thanks for pointing to these issues, BTW).

# Proposal: Xarray flexible indexes refactoring

Current status: https://github.com/pydata/xarray/projects/1

## 1. Data Model

Indexes are used in Xarray to extract data from Xarray objects using coordinate labels instead of using integer array indices. Although the indexes used in an Xarray object can be accessed (or built on-the-fly) via public methods like `to_index()` or properties like `indexes`, those are mainly used internally.

The goal of this project is to make those indexes 1st-class citizens of Xarray's data model. As such, indexes should clearly be separated from Xarray coordinates with the following relationships:

- Index -> Coordinate: one-to-many

> - Coordinate -> Index: one-to-zero-or-one

Reviewer 3: I'm not sure I understand this notation or what it is trying to communicate. Could you make it more verbose.

Author: Yeah I've often seen this terminology in relational database schema modeling that I borrowed from there.

> An index may be built from one or more coordinates. However, each coordinate must relate to one index at most. Additionally, a coordinate may not be tied to any index.

Reviewer 3: Ok, never mind, you said it in the next sentence! 🙃

> The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.

Reviewer: "should" -> "may"

> The order in which multiple coordinates relate to an index should matter. For example, Scikit-Learn's `BallTree` index with the Haversine metric requires providing latitude and longitude values in that specific order. As another example, the order in which levels are defined in a `pandas.MultiIndex` may affect its lexsort depth.

Reviewer 3: Can we make these references links?

> Xarray's current data model is already based on the same index-coordinate relationships. The current data model also implies a one-to-one relationship between a dimension and an index. This one-to-one relationship works as currently it is not possible to perform label-based data selection in Xarray using multi-dimensional coordinates, but since we want to enable this feature in this proposal, we need the dimension-index relationship to evolve towards many-to-many.

Reviewer 3: > Xarray's current data model is already based on the same index-coordinate relationships.

I'm not sure I understand what you're trying to say with this statement. Currently there is a one-to-one relationship between and index and a coordinate, not one-to-many. So it seems inconsistent to claim that the " current data model is already based on the same index-coordinate relationships". What do you mean by "the same"? Be more explicit.

Author: Right, it is currently one-to-one for index->variable, but it may be one-to-many if we consider that virtual coordinates (multi-index) are coordinates. I can clarify this.

> In the example below, we'd like to select data points based on their x, y values and/or on their latitude/longitude positions:

Reviewer 3: Can you be more explicit about how the example relates to what came before? What concept does it illustrate? Many-coordinates-to-one-index?

Author: This example is mainly for illustrating the current limits of a one-to-one relationship for dimension -> index, and why we want this relationship to evolve towards one-to-many.

The many-coordinates-to-one-index relationship is already supported in Xarray with MultiIndex (again considering virtual coordinates), although it is limited too.

Author: I updated the example as well as the paragraph above. This should be clearer now.

```python
>>> da
<xarray.DataArray (x: 2, y: 2)>
array([[5.4, 7.8],
       [6.2, 4.7]])
Coordinates:
  * lon      (x, y) float64 10.2 15.2 12.6 17.6
  * lat      (x, y) float64 40.2 45.6 42.2 47.6
  * x        (x) float64 200.0 400.0
  * y        (y) float64 800.0 1e+03

>>> da.sel(lon=..., lat=...)
>>> da.sel(x=..., y=...)
```

We would need one geographic index for the `lat` and `lon` coordinates and two indexes for the `x` and `y` coordinates, respectively.

## 2. Proposed API changes

### 2.1 Index wrapper classes

Every index that is used to select data from Xarray objects should inherit from a base class, e.g., `XarrayIndex`, that provides some common API. `XarrayIndex` subclasses would generally consist of thin wrappers around existing index classes such as `pandas.Index`, `pandas.MultiIndex`, `scipy.spatial.KDTree`, etc.

There is a variety of features that an xarray index wrapper may or may not support:

> - 1-dimensional vs. 2-dimensional vs. n-dimensional coordinate

Reviewer 3: Can this become more verbose? I can guess, but I don't fully understand what it means to "support" "1-dimensional vs. 2-dimensional vs. n-dimensional coordinate". And how is it different from the next item in the list?

Author: An example is probably more speaking:

- ARGO floats data

```python
da = xr.DataArray(
    data=...,
    dims=('x', 'y'),
    coords={'lat': (('x', 'y'), ...), 'lon': (('x', 'y'), ...)},
    indexes={('lat', 'lon'): GeoSpatialIndex},
)
```

- Ship track data

```python
da = xr.DataArray(
    data=...,
    dims='track',
    coords={'lat': ('track', ...), 'lon': ('track', ...)},
    indexes={'track': GeoSpatialIndex},
)
```

In this example, `GeoSpatialIndex` requires two coordinates exactly (latitude, longitude) but it is compatible with any number of dimensions for those coordinates (maybe except 0-d), like it is currently supported by the geospatial indexes available in `xoak`.

Author: By contrast, a `pandas.MultIndex` may be built from one or more coordinates but they should all have one (the same) dimension.

- built from a single vs multiple coordinate(s)
- in-memory vs. out-of-core (dask) index data/coordinates (vs. other array backends)
- range-based vs. point-wise selection
- exact vs. inexact lookups

Whether or not a `XarrayIndex` subclass supports each of the features listed above should be either declared explicitly via a common API or left to the implementation. An `XarrayIndex` subclass may encapsulate more than one underlying object used to perform the actual indexing. Such "meta" index would typically support a range of features among those mentioned above and would automatically select the optimal index object for a given indexing operation.

Every `XarrayIndex` subclass must at least implement two methods:

> - One `build` method that takes one or more Dataset/DataArray coordinates and that returns the object(s) that will be used for the actual indexing (e.g., `pandas.Index`, `scipy.spatial.KDTree`, etc.)

Reviewer: When would xarray use `build`? How is this different from the `XarrayIndex` constructor?

Alternatively, `XarrayIndex` could do lazy initialization to construct underlying indexes (but that could be an internal implementation detail).

Author: I was thinking of these two advantages of having a `build` method:

- we may want to create an `XarrayIndex` from an existing index object
- a `build` abstract method with a clear signature in the `XarrayIndex` base class might be cleaner than having to specialize `__init__` in subclasses with some implicit requirements?

This could probably makes sense for simple cases... Not for "meta" indexes, though.

More generally, I don't have a clear idea yet on what will be the API of `XarrayIndex` and I guess that it will require multiple iterations during the implementation (I could then move those `build` and `query` to the next item block in the notes).

Reviewer: `build` sounds very generic, but a specialized `from_coords` classmethod for constructing an `XarrayIndex` from just the type would make a lot of sense. Not every index would need a `from_coords` method, only those you want to use in this way.

Author: Yes a `from_coords` class method would be nice! But I'm a bit confused, wouldn't most indexes be built from coordinates? I thought this would rather be the norm, while reusing existing index objects (I mean "underlying" index objects) and extract coordinates would be more occasional..

> - One `query` method that takes label-based indexers as argument and that returns the corresponding position-based indexers.

Reviewer 3: 👍

- `equals()`, `union()` and `intersection()` methods for data alignment (see [Section 2.6](#26-using-indexes-for-data-alignment))
- Xarray coordinate getters (see [Section 2.2.4](#224-implicit-coodinates))
- A method that may return a new index and that will be called when one of the corresponding coordinates is dropped from the Dataset/DataArray (multi-coordinate indexes)
- `encode()`/`decode()` methods that would allow storage-agnostic serialization and fast-path reconstruction of the underlying index object(s) (see [Section 2.8](#28-index-encoding))
- One or more "non-standard" methods or properties that could be leveraged in Xarray 3rd-party extensions like Dataset/DataArray accessors (see [Section 2.7](#27-using-indexes-for-other-purposes))

The `XarrayIndex` API has still to be defined in detail.

Xarray should provide a minimal set of built-in index wrappers (this could be reduced to the indexes currently supported in Xarray, i.e., `pandas.Index` and `pandas.MultiIndex`). Other index wrappers may be implemented in 3rd-party libraries (recommended). The `XarrayIndex` base class should be part of Xarray's public API.

#### 2.1.1 Index discoverability

> For better discoverability of Xarray-compatible indexes, Xarray could provide some mechanism to register new index wrappers, e.g., something lire [xoak's `IndexRegistry`](https://xoak.readthedocs.io/en/latest/_api_generated/xoak.IndexRegistry.html#xoak.IndexRegistry).

Reviewer 3: The [numcodecs registry](https://numcodecs.readthedocs.io/en/stable/registry.html) could be another possible example.

> Additionally (or alternatively), new index wrappers may be registered via entry points like it is already the case for storage backends and maybe other backends (plotting) in the future.

Reviewer 3: ```suggestion
Additionally (or alternatively), new index wrappers may be registered via entry points as is already the case for storage backends and maybe other backends (plotting) in the future.
```

> `XarrayIndex` subclasses may still be used directly when setting new indexes from DataArray/Dataset coordinates.

Reviewer 3: I don't really understand this.

Author: It means that both

```python
ds.set_index({('lat', 'lon'): xr.index_registry.geospatial_index})
```

and 

```python
from xarray import XarrayIndex

class MyGeoSpatialIndex(XarrayIndex):
    ...

ds.set_index({('lat', 'lon'): MyGeoSpatialIndex})
```

would be supported.

### 2.2 Explicit vs. implicit index creation

#### 2.2.1 Dataset/DataArray's `indexes` constructor argument

The new `indexes` argument of Dataset/DataArray constructors may be used to specify which kind of index to bind to which coordinate(s). It would consist of a mapping where, for each item, the key is one coordinate name (or a sequence of coordinate names) that must be given in `coords` and the value is the type of the index to build from this (these) coordinate(s):

```python
>>> da = xr.DataArray(
...     data=[[275.2, 273.5], [270.8, 278.6]],
...     dims=('x', 'y'),
...     coords={
...         'lat': (('x', 'y'), [[45.6, 46.5], [50.2, 51.6]]),
...         'lon': (('x', 'y'), [[5.7, 10.5], [6.2, 12.8]]),
...     },
...     indexes={('lat', 'lon'): SpatialIndex},
... )
<xarray.DataArray (x: 2, y: 2)>
array([[275.2, 273.5],
       [270.8, 278.6]])
Coordinates:
  * lat      (x, y) float64 45.6 46.5 50.2 51.6
  * lon      (x, y) float64 5.7 10.5 6.2 12.8
```

More formally, `indexes` would accept `Mapping[CoordinateNames, IndexSpec]` where:

- `CoordinateNames = Union[CoordinateName, Tuple[CoordinateName, ...]]` and `CoordinateName = Hashable`

> - `IndexSpec = Union[Type[XarrayIndex], Tuple[XarrayIndex, Dict[str, Any]], XarrayIndex]`, so that index instances or index classes + build options could be also passed

Reviewer: I'm not sure I exactly understand what each of these options would correspond to. I guess `XarrayIndex` is the only essential one, and the others are for user convenience?

Author: There's a mistake: `Tuple[XarrayIndex, Dict[str, Any]]` ->  `Tuple[Type[XarrayIndex], Dict[str, Any]]`

Some examples:

- `Type[XarrayIndex]`

```python
da = xr.DataArray(
    data=...,
    dims=('x', 'y'),
    coords={'lat': (('x', 'y'), ...), 'lon': (('x', 'y'), ...)},
    indexes={('lat', 'lon'): SpatialIndex},
)
```

- `Tuple[Type[XarrayIndex], Dict[str, Any]]`

```python
>>> da = xr.DataArray(
    data=...,
    dims=('x', 'y'),
    coords={'lat': (('x', 'y'), ...), 'lon': (('x', 'y'), ...)},
    indexes={('lat', 'lon'): (SpatialIndex, {'kdtree_leafsize': 20})},
)
```

- `XarrayIndex`

```python
midx = pd.MultiIndex.from_arrays(..., names=['lvl1' , 'lvl2'])

coords_dict, index_dict = create_coords_from_index(midx, dims='x')

# index_dict == {('lvl1', lvl2'): midx}

da =  xr.DataArray(..., dims='x', coords=coords_dict, indexes=index_dict)
```

Option A looks simple and elegant but significantly departs from the current signature. Option B is more consistent with the Dataset/DataArray constructor signature proposed in the previous section and would be easier to adopt in parallel with the current signature that we could still support through some depreciation cycle.

The `append` parameter of the current `.set_index()` is specific to `pandas.MultiIndex`. With option B we could still support it, although we might want to either drop it or move it to the index construction options in the future.

#### 2.2.3 Implicit default indexes

In general explicit index creation should be preferred over implicit index creation. However, there is a majority of cases where basic `pandas.Index` objects could be built and used as indexes for 1-dimensional coordinates. For convenience, Xarray should automatically build such indexes for the coordinates where no index has been explicitly assigned in the Dataset/DataArray constructor or when indexes have been reset / dropped.

For which coordinates?

- A. only 1D coordinates with a name matching their dimension name

> - B. all 1D coordinates

Reviewer 3: :+1: to all 1D coordinates. I call `swap_dims` often just to make it easier to select based on different 1D coordinates. This would eliminate that. But could get expensive for datasets with lots of long non-dimension coordinates.

Also, automatically creating indexes could trigger eager loading of coordinates that would otherwise be lazy. For some datasets this would be catastrophic for memory usage (related to #1650).

Author: Once we have some kind of out-of-core index in Xarray (as suggested in #1650) I guess we could safely extend the implicit index creation to all 1D coordinates, but in the meantime it's probably wiser to limit that to dimension coordinates for the eager loading issues that you mention. Instead of calling `swap_dims`, you'll just need one explicit call to `set_index`.

 array(['a', 'b'], dtype=object),
 'lvl2': <xarray.Variable (x: 2)>
 array([0, 1])}
>>> index_dict
{('lvl1', 'lvl2'): midx}
>>> xr.DataArray([1.0, 2.0], dims='x', coords=coords_dict, indexes=index_dict)
<xarray.DataArray (x: 2)>
array([1., 2.])
Coordinates:
    x        (x) object ('a', 0) ('b', 1)
  * lvl1     (x) object 'a' 'b'
  * lvl2     (x) int64 0 1
```

### 2.2.5 Immutable indexes

Some underlying indexes might be mutable (e.g., a tree-based index structure that allows dynamic addition of data points) while other indexes like `pandas.Index` aren't. To keep things simple, it is probably better to continue considering all indexes in Xarray as immutable (as well as their corresponding coordinates, see [Section 2.4.1](#241-mutable-coordinates)).

### 2.3 Index access

#### 2.3.1 Dataset/DataArray's `indexes` property

The `indexes` property would allow easy access to all the indexes used in a Dataset/DataArray. There may be different options for the type returned:

A. `Dict[CoordinateName, XarrayIndex]`: keys are coordinate names and values may be duplicated
B. `Dict[CoordinateNames, XarrayIndex]`: keys may represent one or more coordinate names and values are unique

> Option A allows easy index look-up by coordinate name, while option B allows easy iteration through all the indexes. Option A may be more useful than option B for many tasks and may also be less ambiguous if more complex hashable types than `str` are used for `CoordinateName`.

Reviewer: I think I mentioned this before, but I think option A is much better.

Here's an example implementation of the two "hard" operations with each data format:

Iterate over all indexes with A:
```python
def iter_all(indexes):
    output_ids = set()
    for index in indexes.values():
        index_id = id(index)
        if index_id not in output_ids:
            output_ids.add(index_id)
            yield index
```

Look-up by coordinate name with option B (basically needs to build up the "option A" representation internally):
```python
def lookup_by_name(indexes, coord_name):
    coord_name_to_index = {}
    for names, index in indexes.items():
        # this would be even worse if we needed to check whether
        # names is a tuple or a single name
        for name in names:
            coord_name_to_index[name] = index
    return coord_name_to_index[coord_name]
```

Comparing the run-time by computational complexity:

- Look-up by coordinate name: `O(1)` with option A vs `O(number_of_indexed_coordinates)` with Option B
- Iterate over all unique indexes: `O(number_of_indexed_coordinates)` with option A vs `O(number_of_unique_indexes)`

I expect "look-up by coordinate name" would be most common in performance limited situation (e.g., in indexing operations), and there's a much more meaningful speed-up there. In most cases, I would guess `number_of_indexed_coordinates` is not much larger than `number_of_unique_indexes`.

There's also the whole separate issue of option B potentially allowing both strings and tuples of strings as keys. Even if we do option B, we should only use tuples to avoid potential ambiguity and needing to handle both cases every time the index is used.

Author: Agreed, you already convinced me before that option A is better (not sure that performance is really critical here for only a few coordinates / indexes per dataset, but in general I agree with your arguments in favor of A).

I intentionally left the options in the notes (along with some explanation of why one an option is better than the other), but if we all agree then I can remove option B here.

> #### 2.3.2 Additional Dataset/DataArray's properties or methods

Reviewer 3: ```suggestion
#### 2.3.2 Additional Dataset/DataArray properties or methods
```

#### 2.4.2 New Dataset/DataArray with updated coordinates

Xarray provides a variety of Dataset/DataArray operations affecting the coordinates and where simply dropping the index(es) is not desirable. For example:

- multi-coordinate indexes could be reduced to single coordinate indexes
  - like in `.reset_index()` or `.sel()` applied on a subset of the levels of a `pandas.MultiIndex` and that internally call `MultiIndex.droplevel` and `MultiIndex.get_loc_level`, respectively
- indexes may be indexed themselves
  - like `pandas.Index` implements `__getitem__()`
  - when indexing their corresponding coordinate(s), e.g., via `.sel()` or `.isel()`, those indexes should be indexed too
  - this might not be supported by all Xarray indexes, though
- some indexes that can't be indexed could still be automatically (re)built in the new Dataset/DataArray
  - like for example building a new `KDTree` index from the selection of a subset of an initial collection of data points
  - this is not always desirable, though, as indexes may be expensive to build

>   - a more reasonable option would be to explicitly re-build the index, e.g., using `.set_index()`

Reviewer: +1

- there should be the possibility to provide additional options to the indexes that support specific selection features (e.g., Scikit-learn's `BallTree`'s `dualtree` query option to boost performance).
  - the best API is not trivial here, since `.sel()` may accept indexers passed to several indexes (which should still be supported for convenience and compatibility), and indexes may have similar options with different semantics
  - we could introduce a new parameter like `index_options: Dict[XarrayIndex, Dict[str, Any]]` to pass options grouped by index
- the `method` and `tolerance` parameters are specific to `pandas.Index` and would not be supported by all indexes: probably best is to eventually pass those arguments as `index_options`
- the list valid indexer types might be extended in order to support new ways of indexing data, e.g., unordered selection of all points within a given range
  - alternatively, we could reuse existing indexer types with different semantics depending on the index, e.g., using `slice(min, max, None)` for unordered range selection

> With the new data model proposed here, once ambiguous situation may occur when indexers are given for several coordinates that share the same dimension but not the same index, e.g., from the example in [Section 1](#1-data-model):

Reviewer 3: ```suggestion
With the new data model proposed here, an ambiguous situation may occur when indexers are given for several coordinates that share the same dimension but not the same index, e.g., from the example in [Section 1](#1-data-model):
```

### 2.6 Using indexes for data alignment

Another main use if indexes is data alignment in various operations. Some considerations regarding alignment and flexible indexes:

- support for alignment should probably be optional for an `XarrayIndex` subclass.
  - like `pandas.Index`, the index wrapper classes that support it should implement `.equals()`, `.union()` and/or `.intersection()`
  - support might be partial if that makes sense (outer, inner, left, right, exact...).
  - index equality might involve more than just the labels: for example a spatial index might be used to check if the coordinate system (CRS) is identical for two sets of coordinates
  - some indexes might implement inexact alignment, like in [#4489](https://github.com/pydata/xarray/pull/4489) or a `KDTree` index that selects nearest-neighbors within a given tolerance
  - alignment may be "multi-dimensional", i.e., the `KDTree` example above vs. dimensions aligned independently of each other

> - we need to decide what to do when one dimension has more than one index that supports alignment

Reviewer 3: Here and elsewhere it sounds like it could be useful to define an index precedence order.

Author: Yes this could be a solution to all situations where the index to pick is ambiguous (although it is not yet clear to me how often such situation could happen). This would introduce more complexity, though:

- we need to extend the API so that users can define this order
- we need to ensure strict precedence order
- does it make sense to use the same precedence order for all operations? (e.g., alignment vs. selection)

- we need to decide what to do when one dimension has one or more index(es) but none support alignment
  - either we raise or we fail back (silently) to alignment based on dimension size
- for inexact alignment, the tolerance threshold might be given when building the index and/or when performing the alignment
- are there cases where we want a specific index to perform alignment and another index to perform selection?
  - it would be tricky to support that unless we allow multiple indexes per coordinate
  - alternatively, underlying indexes could be picked internally in a "meta" index for one operation or another, although the risk is to eventually have to deal with an explosion of index wrapper classes with different meta indexes for each combination that we'd like to use.

### 2.7 Using indexes for other purposes

Xarray also provides a number of Dataset/DataArray methods where indexes are used in various ways, e.g.,

- `resample` (`CFTimeIndex` and a `DatetimeIntervalIndex`)
- `DatetimeAccessor` & `TimedeltaAccessor` properties (`CFTimeIndex` and a `DatetimeIntervalIndex`)
- `interp` & `interpolate_na`,

>    - with `IntervalIndex`, these become regridding operations. Should we support hooks for these operations?

Reviewer 3: I think indexes should optionally support interpolation.

Author: That would certainly be useful.

- `differentiate`, `integrate`, `polyfit`
   - raise an error if not a "simple" 1D index?
- `pad`
- `coarsen` has to make choices about output index labels.
- `sortby`
- `stack`/`unstack`
- plotting
    - `plot.pcolormesh` "infers" interval breaks along axes, which are really inferred `bounds` for the appropriate indexes.
    - `plot.step` again uses `bounds`. In fact, we may even want `step` to be the default 1D plotting function if the axis has `bounds` attached.

It would be reasonable to first restrict those methods to the indexes that are currently available in Xarray, and maybe extend the `XarrayIndex` API later upon request when the opportunity arises.

Conversely, nothing should prevent implementing "non-standard" API in 3rd-party `XarrayIndex` subclasses that could be used in DataArray/Dataset extensions (accessors). For example, we might want to reuse a `KDTree` index to compute k-nearest neighbors (returning a DataArray/Dataset with a new dimension) and/or the distances to the nearest neighbors (returning a DataArray/Dataset with a new data variable).

### 2.8 Index encoding

> Indexes don't need to be directly serializable since we could (re)build them from their corresponding coordinate(s). However, we may take advantage that some indexes could be encoded/decoded to/from a set of arrays that would allow optimized reconstruction and/or storage, e.g.,

Reviewer 3: ```suggestion
Indexes don't need to be directly serializable since we could (re)build them from their corresponding coordinate(s). However, it would be useful if some indexes could be encoded/decoded to/from a set of arrays that would allow optimized reconstruction and/or storage, e.g.,
```

array([[5.4, 7.8],
       [6.2, 4.7]])
Coordinates:
  * lon      (x, y) float64 10.2 15.2 12.6 17.6
  * lat      (x, y) float64 40.2 45.6 42.2 47.6
  * x        (x) float64 200.0 400.0
  * y        (y) float64 800.0 1e+03
Indexes:
  lat, lon     <SpatialIndex coords=(lat, lon) dims=(x, y)>
  x            <PandasIndexWrapper>
  y            <PandasIndexWrapper>
```

To keep the `repr` compact, we could:

- consolidate entries that map to the same index object, and have an short inline repr for `XarrayIndex` object
- collapse the index section by default in the HTML `repr`
- maybe omit all trivial indexes for 1D coordinates that match the dimension name

## 4. `IndexVariable`

`IndexVariable` is currently used to wrap an `pandas.Index` as a variable. Although the main goal of this refactoring is to decouple indexes and variables (coordinates), `IndexVariable` could still be useful to:

- ensure that all coordinates with an index are immutable (see [Section 2.4.1](#241-mutable-coordinates))
  - do not set values directly, do not (re)chunk (even though it may be already chunked), do not load, do not convert to sparse/dense, etc.
- directly reuse index's data when that's possible

>   - in the case of a `pandas.Index`, it makes little sense to have duplicate data (e.g., as a NumPy array) for its corresponding coordinate

Reviewer: In the long term, I think it would probably be ideal to eliminate `IndexVariable`, which is just one more thing that can get out of sync when setting/resetting a index.

Instead, we could ensure that the `data` associated with a `Variable` is immutable by writing a custom array wrapper class like `CopyOnWriteArray` that raises an error from `__setitem__`.
